File: android/feature/events/src/main/java/com/creapolis/solennix/feature/events/viewmodel/EventChecklistViewModel.kt
package com.creapolis.solennix.feature.events.viewmodel

import android.content.Context
import android.content.SharedPreferences
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.creapolis.solennix.core.data.repository.EventRepository
import com.creapolis.solennix.core.data.repository.ProductRepository
import com.creapolis.solennix.core.model.SupplySource
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ChecklistItem(
    val id: String,
    val name: String,
    val quantity: Double,
    val unit: String,
    val section: ChecklistSection,
    val unitCost: Double = 0.0
)

enum class ChecklistSection {
    EQUIPMENT,
    STOCK,
    PURCHASE
}

data class EventChecklistUiState(
    val eventId: String = "",
    val eventName: String = "",
    val eventDate: String = "",
    val items: List<ChecklistItem> = emptyList(),
    val checkedIds: Set<String> = emptySet(),
    val isLoading: Boolean = true,
    val error: String? = null
) {
    val progress: Float
        get() = if (items.isEmpty()) 0f else checkedIds.size.toFloat() / items.size

    val equipmentItems: List<ChecklistItem>
        get() = items.filter { it.section == ChecklistSection.EQUIPMENT }

    val stockItems: List<ChecklistItem>
        get() = items.filter { it.section == ChecklistSection.STOCK }

    val purchaseItems: List<ChecklistItem>
        get() = items.filter { it.section == ChecklistSection.PURCHASE }
}

@HiltViewModel
class EventChecklistViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val eventRepository: EventRepository,
    private val productRepository: ProductRepository,
    @ApplicationContext private val context: Context
) : ViewModel() {

    private val eventId: String = savedStateHandle["eventId"] ?: ""

    // Checklist progress is persisted per-event; event IDs end up as preference
    // keys. Keep the file encrypted so a rooted device or backup extraction
    // cannot enumerate the user's event IDs from plain preference files.
    private val prefs: SharedPreferences = run {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            context,
            "checklist_prefs_encrypted",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    private val _uiState = MutableStateFlow(EventChecklistUiState(eventId = eventId))
    val uiState: StateFlow<EventChecklistUiState> = _uiState.asStateFlow()

    init {
        loadChecklist()
        loadPersistedCheckedState()
    }

    private fun loadChecklist() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            try {
                val event = eventRepository.getEvent(eventId)
                if (event == null) {
                    _uiState.update { it.copy(isLoading = false, error = "Evento no encontrado") }
                    return@launch
                }

                // Fetch event-specific data concurrently
                val productsDeferred = async { eventRepository.getEventProductsFromApi(eventId) }
                val equipmentDeferred = async { eventRepository.getEventEquipmentFromApi(eventId) }
                val suppliesDeferred = async { eventRepository.getEventSuppliesFromApi(eventId) }

                val products = productsDeferred.await()
                val equipment = equipmentDeferred.await()
                val supplies = suppliesDeferred.await()

                val checklistItems = mutableListOf<ChecklistItem>()

                // 1. Equipment items from event assignments
                for (item in equipment) {
                    checklistItems.add(
                        ChecklistItem(
                            id = "eq_${item.id}",
                            name = item.equipmentName ?: "Equipo",
                            quantity = item.quantity.toDouble(),
                            unit = item.unit ?: "",
                            section = ChecklistSection.EQUIPMENT
                        )
                    )
                }

                // 2. Product ingredients with bringToEvent=true, aggregated by inventoryId
                data class IngredientAgg(var name: String, var quantity: Double, var unit: String)
                val ingredientMap = mutableMapOf<String, IngredientAgg>()

                for (product in products) {
                    try {
                        val ingredients = productRepository.getProductIngredients(product.productId)
                        for (ingredient in ingredients) {
                            if (ingredient.bringToEvent != true) continue
                            val key = ingredient.inventoryId
                            val totalQty = ingredient.quantityRequired * product.quantity
                            val existing = ingredientMap[key]
                            if (existing != null) {
                                existing.quantity += totalQty
                            } else {
                                ingredientMap[key] = IngredientAgg(
                                    name = ingredient.ingredientName ?: "Ingrediente",
                                    quantity = totalQty,
                                    unit = ingredient.unit ?: ""
                                )
                            }
                        }
                    } catch (_: Exception) {
                        // Skip products whose ingredients can't be fetched
                    }
                }

                for ((inventoryId, info) in ingredientMap) {
                    checklistItems.add(
                        ChecklistItem(
                            id = "ing_$inventoryId",
                            name = info.name,
                            quantity = info.quantity,
                            unit = info.unit,
                            section = ChecklistSection.STOCK
                        )
                    )
                }

                // 3. Supply items — section determined by source field
                for (supply in supplies) {
                    val section = if (supply.source == SupplySource.STOCK) {
                        ChecklistSection.STOCK
                    } else {
                        ChecklistSection.PURCHASE
                    }
                    checklistItems.add(
                        ChecklistItem(
                            id = "sup_${supply.id}",
                            name = supply.supplyName ?: "Insumo",
                            quantity = supply.quantity,
                            unit = supply.unit ?: "",
                            section = section,
                            unitCost = supply.unitCost
                        )
                    )
                }

                _uiState.update {
                    it.copy(
                        isLoading = false,
                        eventName = event.serviceType,
                        eventDate = event.eventDate,
                        items = checklistItems
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }

    private fun loadPersistedCheckedState() {
        val key = "checklist_$eventId"
        val checkedIds = prefs.getStringSet(key, emptySet()) ?: emptySet()
        _uiState.update { it.copy(checkedIds = checkedIds) }
    }

    private fun persistCheckedState(checkedIds: Set<String>) {
        val key = "checklist_$eventId"
        prefs.edit().putStringSet(key, checkedIds).apply()
    }

    fun toggleItem(itemId: String) {
        val currentChecked = _uiState.value.checkedIds
        val newChecked = if (currentChecked.contains(itemId)) {
            currentChecked - itemId
        } else {
            currentChecked + itemId
        }

        _uiState.update { it.copy(checkedIds = newChecked) }
        persistCheckedState(newChecked)
    }

    fun checkAllInSection(section: ChecklistSection) {
        val sectionItems = _uiState.value.items.filter { it.section == section }
        val sectionIds = sectionItems.map { it.id }.toSet()
        val newChecked = _uiState.value.checkedIds + sectionIds

        _uiState.update { it.copy(checkedIds = newChecked) }
        persistCheckedState(newChecked)
    }

    fun uncheckAllInSection(section: ChecklistSection) {
        val sectionItems = _uiState.value.items.filter { it.section == section }
        val sectionIds = sectionItems.map { it.id }.toSet()
        val newChecked = _uiState.value.checkedIds - sectionIds

        _uiState.update { it.copy(checkedIds = newChecked) }
        persistCheckedState(newChecked)
    }

    fun resetChecklist() {
        _uiState.update { it.copy(checkedIds = emptySet()) }
        val key = "checklist_$eventId"
        prefs.edit().remove(key).apply()
    }
}

File: ios/Packages/SolennixFeatures/Sources/SolennixFeatures/Events/ViewModels/EventChecklistViewModel.swift
import Foundation
import Observation
import SolennixCore
import SolennixNetwork

// MARK: - Checklist Section

public enum ChecklistSection: String, Sendable, Hashable {
    case equipment
    case stock
    case purchase
}

// MARK: - Checklist Item

public struct ChecklistItem: Identifiable, Sendable, Hashable {
    public let id: String
    public let name: String
    public let quantity: Double
    public let unit: String?
    public let section: ChecklistSection
}

// MARK: - Event Checklist View Model

@MainActor
@Observable
public final class EventChecklistViewModel {

    // MARK: - Properties

    public var items: [ChecklistItem] = []
    public var checkedIds: Set<String> = []
    public var isLoading: Bool = false
    public var errorMessage: String?

    private var eventId: String = ""

    // MARK: - Dependencies

    private let apiClient: APIClient

    // MARK: - Init

    public init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    // MARK: - Computed

    public var progress: Double {
        guard !items.isEmpty else { return 0 }
        return Double(checkedIds.count) / Double(items.count)
    }

    public var equipmentItems: [ChecklistItem] {
        items.filter { $0.section == .equipment }
    }

    public var stockItems: [ChecklistItem] {
        items.filter { $0.section == .stock }
    }

    public var purchaseItems: [ChecklistItem] {
        items.filter { $0.section == .purchase }
    }

    // MARK: - Load Checklist

    @MainActor
    public func loadChecklist(eventId: String) async {
        self.eventId = eventId
        isLoading = true
        errorMessage = nil

        do {
            // Fetch event products, equipment, and supplies concurrently
            async let productsResult: [EventProduct] = apiClient.get(Endpoint.eventProducts(eventId))
            async let equipmentResult: [EventEquipment] = apiClient.get(Endpoint.eventEquipment(eventId))
            async let suppliesResult: [EventSupply] = apiClient.get(Endpoint.eventSupplies(eventId))

            let products = try await productsResult
            let equipment = try await equipmentResult
            let supplies = try await suppliesResult

            var checklistItems: [ChecklistItem] = []

            // Equipment items
            for item in equipment {
                checklistItems.append(ChecklistItem(
                    id: "eq_\(item.id)",
                    name: item.equipmentName ?? "Equipo",
                    quantity: Double(item.quantity),
                    unit: item.unit,
                    section: .equipment
                ))
            }

            // Fetch product ingredients for aggregation
            var ingredientMap: [String: (name: String, quantity: Double, unit: String?)] = [:]

            for product in products {
                do {
                    let ingredients: [ProductIngredient] = try await apiClient.get(
                        Endpoint.productIngredients(product.productId)
                    )

                    for ingredient in ingredients {
                        guard ingredient.bringToEvent == true else { continue }
                        let key = ingredient.inventoryId
                        let totalQty = ingredient.quantityRequired * Double(product.quantity)

                        if var existing = ingredientMap[key] {
                            existing.quantity += totalQty
                            ingredientMap[key] = existing
                        } else {
                            ingredientMap[key] = (
                                name: ingredient.ingredientName ?? "Ingrediente",
                                quantity: totalQty,
                                unit: ingredient.unit
                            )
                        }
                    }
                } catch {
                    // Skip products whose ingredients can't be fetched
                }
            }

            // Add aggregated ingredients as stock items
            for (inventoryId, info) in ingredientMap {
                checklistItems.append(ChecklistItem(
                    id: "ing_\(inventoryId)",
                    name: info.name,
                    quantity: info.quantity,
                    unit: info.unit,
                    section: .stock
                ))
            }

            // Supply items
            for supply in supplies {
                let section: ChecklistSection = supply.source == .stock ? .stock : .purchase
                checklistItems.append(ChecklistItem(
                    id: "sup_\(supply.id)",
                    name: supply.supplyName ?? "Insumo",
                    quantity: supply.quantity,
                    unit: supply.unit,
                    section: section
                ))
            }

            items = checklistItems

            // Load checked state from UserDefaults
            loadCheckedState()
        } catch {
            if let apiError = error as? APIError {
                errorMessage = apiError.errorDescription ?? "Error cargando el checklist"
            } else {
                errorMessage = "Error cargando el checklist"
            }
        }

        isLoading = false
    }

    // MARK: - Toggle Item

    public func toggleItem(id: String) {
        if checkedIds.contains(id) {
            checkedIds.remove(id)
        } else {
            checkedIds.insert(id)
        }
        saveCheckedState()
    }

    // MARK: - Persistence

    private func loadCheckedState() {
        let key = "event_checklist_\(eventId)"
        if let saved = UserDefaults.standard.array(forKey: key) as? [String] {
            checkedIds = Set(saved)
        } else {
            checkedIds = []
        }
    }

    private func saveCheckedState() {
        let key = "event_checklist_\(eventId)"
        UserDefaults.standard.set(Array(checkedIds), forKey: key)
    }
}
